examples/fabric-admin/commands/pairing/DeviceSynchronization.cpp

```diff
@@ -0,0 +1,186 @@
+        break;
+    case Clusters::BasicInformation::Attributes::VendorName::Id:
+        mCurrentDeviceData.has_vendor_name =
+            SuccessOrLog(data->GetString(mCurrentDeviceData.vendor_name, sizeof(mCurrentDeviceData.vendor_name)), "VendorName");
+        break;
+    case Clusters::BasicInformation::Attributes::VendorID::Id:
+        mCurrentDeviceData.has_vendor_id = SuccessOrLog(data->Get(mCurrentDeviceData.vendor_id), "VendorID");
+        break;
+    case Clusters::BasicInformation::Attributes::ProductName::Id:
+        mCurrentDeviceData.has_product_name =
+            SuccessOrLog(data->GetString(mCurrentDeviceData.product_name, sizeof(mCurrentDeviceData.product_name)), "ProductName");
+        break;
+    case Clusters::BasicInformation::Attributes::ProductID::Id:
+        mCurrentDeviceData.has_product_id = SuccessOrLog(data->Get(mCurrentDeviceData.product_id), "ProductID");
+        break;
+    case Clusters::BasicInformation::Attributes::NodeLabel::Id:
+        mCurrentDeviceData.has_node_label =
+            SuccessOrLog(data->GetString(mCurrentDeviceData.node_label, sizeof(mCurrentDeviceData.node_label)), "NodeLabel");
+        break;
+    case Clusters::BasicInformation::Attributes::HardwareVersion::Id:
+        mCurrentDeviceData.has_hardware_version = SuccessOrLog(data->Get(mCurrentDeviceData.hardware_version), "HardwareVersion");
+        break;
+    case Clusters::BasicInformation::Attributes::HardwareVersionString::Id:
+        mCurrentDeviceData.has_hardware_version_string = SuccessOrLog(
+            data->GetString(mCurrentDeviceData.hardware_version_string, sizeof(mCurrentDeviceData.hardware_version_string)),
+            "HardwareVersionString");
+        break;
+    case Clusters::BasicInformation::Attributes::SoftwareVersion::Id:
+        mCurrentDeviceData.has_software_version = SuccessOrLog(data->Get(mCurrentDeviceData.software_version), "HardwareVersion");
+        break;
+    case Clusters::BasicInformation::Attributes::SoftwareVersionString::Id:
+        mCurrentDeviceData.has_software_version_string = SuccessOrLog(
+            data->GetString(mCurrentDeviceData.software_version_string, sizeof(mCurrentDeviceData.software_version_string)),
+            "SoftwareVersionString");
+        break;
+    default:
+        break;
+    }
+}
+
+void DeviceSynchronizer::OnReportEnd()
+{
+    // Report end is at the end of all attributes (success)
+#if defined(PW_RPC_ENABLED)
+    AddSynchronizedDevice(mCurrentDeviceData);
+#else
+    ChipLogError(NotSpecified, "Cannot synchronize device with fabric bridge: RPC not enabled");
+#endif
+    mDeviceSyncInProcess = false;
+}
+
+void DeviceSynchronizer::OnDone(chip::app::ReadClient * apReadClient)
+{
+    // Nothing to do: error reported on OnError or report ended called.
+}
+
+void DeviceSynchronizer::OnError(CHIP_ERROR error)
+{
+    ChipLogProgress(NotSpecified, "Error fetching device data: %" CHIP_ERROR_FORMAT, error.Format());
+    mDeviceSyncInProcess = false;
```

Comment:
Same thing here, we should only set this to false in OnDone

https://github.com/project-chip/connectedhomeip/blob/master/src/app/ReadClient.h#L192-L207

Reply:
Update to OnDone and also on sending request error. Kind of brittle :(